### ISSUE TITLE
Passing arbitrary parameters to the ID assertion endpoint

### ISSUE THREAD

**Issue by cbiesinger** [fedidcg/FedCM#556]
(this has been split out of #477 )

Also, there is currently no general-purpose way to pass information from the RP to the ID assertion endpoint. It is already possible to pass arbitrary data through the “nonce” or the “client_id” field; however, this is inconvenient and unstructured. An explicit way to pass key/value pairs would be better.

**Comment by cbiesinger** [comment]
## Proposal

We are proposing letting the RP specify additional fields that will get sent to the IdP:


```
partial dictionary IdentityProviderConfig {
  // A list of strings that represent what the Relying Party needs
  // from the Identity Provider.
  sequence<USVString> scope;
  // The types of tokens that the RP wants to get back from the IdP.
  // More than one can be requested, e.g. an id_token and an access code
  // token:
  // https://openid.net/specs/openid-connect-core-1_0.html
  // #code-id_token-tokenExample
  sequence<USVString> responseType;
  // A map of key-value pairs that are opaque to the browser and only
  // passed to the IdP after the user's acknowledgement.
  record<USVString, USVString> params;
};
```


An RP could use them like this: \
`let {token} = await navigator.credentials.get({`


```
  identity: {
    providers: [{
      // A newly introduced "scope" attribute, a list of strings
      // Defaults to "name email photo"
      scope: [
        "drive.readonly",
        "calendar.readonly",
      ],


      clientId: "1234",
      nonce: "234234",
      loginHint: "previous@user.com",
      configURL: "https://idp.example/fedcm.json",
      responseType: ["id_token"],
      // A string with parameters that need to be passed from the
      // RP to the IdP but that don't really play any role with
      // the browser.
      params: {
        "IDP_SPECIFIC_PARAM": "1",
        "foo": "BAR",
        "ETC": "MOAR",
      }
    },
  }
  // If possible, return without prompting the user, if not possible
  // prompt the user.
  mediation: "optional",
});
```


These parameters would be used in two ways:



* They will be sent to the ID assertion endpoint (that is, they will only be shared with the IdP after user interaction)
* If a non-default scope will be used, the browser will not show the sharing disclosure text (i.e. the “To continue, idp.example will share your name, email address and profile picture with this site. See this site’s <span style="text-decoration:underline;">privacy policy</span> and <span style="text-decoration:underline;">terms of service</span>.” text) \
 \
That is, the account chooser will behave exactly as if this was a returning user.

For example, the ID assertion request body might look like this: \
`POST /fedcm_assertion_endpoint HTTP/1.1`


```
Host: idp.example
Origin: https://rp.example/
Content-Type: application/x-www-form-urlencoded
Cookie: 0x23223
Sec-Fetch-Dest: webidentity

account_id=123&client_id=client1234&nonce=234234&disclosure_text_shown=false&scope=google-drive.readonly&reponseType=id_token&scope=drive.readonly+calendar.readonly&IDP_SPECIFIC_PARAM=1&foo=BAR&ETC=MOAR
```


This follows the [OIDC standard](https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims) for specifying scopes (naming of “scope” and separating them with spaces, here URL-encoded using +).

If necessary, the IdP can use the continue_on API to gather the user’s permission for accessing the requested scopes.

An RP could silently request a refresh token like this, provided that the user had previously permitted signing in to this RP with this IDP:


```
let result = await navigator.credentials.get({
  // Returns a result without any UI in case it is possible
  // (specifically, only after the user has accepted the FedCM
  // prompt in the past). Otherwise, returns an error.
  mediation: "silent",
  identity: {
    providers: [{
      // newly introduced "scope" attribute, a list of strings
      scope: [
        "google-drive.readonly"
      ],
      configURL: "https://idp.example/fedcm.json",
      clientId: "1234",
    }]
  }
});
```



## Considerations

Specifying `scope` allows the RP to disable the sharing disclosure text even for new users, and the browser cannot guarantee that the IDP will itself show an equivalent disclosure. However, a well-behaved IDP will ensure that such a disclosure is shown (using the `disclosure_text_shown` if necessary, and use a `continue_on` URL to gather user permission). Conversely, a non-well-behaved IDP can just send an `approved_clients` list to suppress showing the disclosure text in the first place.

Because the `responseType` parameter is treated as opaque data by the browser and sent as-is to the  IdP, it is not strictly speaking necessary and could be replaced with a `params` entry that the RP and IdP agree on. However, it allows more explicit semantics.

As suggested above, custom parameters are added to the response body as-is. That means they can conflict with built-in keys such as account_id. We could disallow keys that conflict with built-in keys, or prepend a special prefix to all parameters.

This proposal works best in conjunction with the continuation API proposal.

**Comment by cbiesinger** [comment]
I have split out the scope parameter to #559 because that seems more controversial.

**Comment by cbiesinger** [comment]
The discussion in today's fedid CG meeting was in favor of adding params, but was skeptical about responseType. I think it's best to remove that field from this proposal.

**Comment by panva** [comment]
Some OAuth 2.0 registered extension parameters, e.g. `resource` can be passed to the authorization endpoint multiple times.

To that end with a definition like so 

```
  record<USVString, USVString> params;
```

cannot be used to achieve a application/x-www-form-urlencoded encoded body which an OAuth 2.0 authorization server understands

```
resource=urn:example:1&resource=urn:example:2
```

**Comment by panva** [comment]
```
partial dictionary IdentityProviderConfig {
  sequence<USVString> responseType;
  record<USVString, USVString> params;
};
```

`sequence<USVString> responseType;` is redundant if I can pass any parameter, to that end so would the existing `nonce`.

**Comment by samuelgoto** [comment]
> sequence<USVString> responseType; is redundant if I can pass any parameter, to that end so would the existing nonce and the proposed #559 Scope API.

Yep. @cbiesinger is removing `responseType` from this proposal. See above:

https://github.com/fedidcg/FedCM/issues/556#issuecomment-2072677672

**Comment by cbiesinger** [comment]
The scope API is partially redundant, but it also affects the browser UI (which is why it is a separate proposal).

You are correct about nonce, but I am not sure we will be able to remove it for backwards compatibility reasons.

**Comment by cbiesinger** [comment]
@panva thanks for the information. however I am not sure that exact compatibility with existing endpoints is a goal

**Comment by panva** [comment]
> @panva thanks for the information. however I am not sure that exact compatibility with existing endpoints is a goal

I am sure that was not my point. I simply ask to use a type that allows prior art that is deployed in the wild to be used.

**Comment by obfuscoder** [comment]
Would it be possible to allow for an array of strings for param values? This could be constructed as a union to values as strings abut also array of strings. This would allow for the multiple resource example given by @panva

**Comment by samuelgoto** [comment]
> Would it be possible to allow for an array of strings for param values? 

I think the proposal is to accept an `object`, which is a set of key-value pairs (where the keys and values are strings), so I think the answer is "yes".

**Comment by cbiesinger** [comment]
As of tomorrow's canary, responseType will no longer be supported in Chrome

Regarding arrays of strings, it is certainly doable if we decide to support it.

**Comment by domenic** [comment]
To be a bit clearer on the usual API shape for accepting these sorts of things: it's

```webidl
sequence<sequence<USVString>> or record<USVString, USVString>
```

with extra processing on the first variant. See https://fetch.spec.whatwg.org/#concept-headers-fill for the processing algorithm.

This allows the formats

```js
{
  "key": "value",
  "key2": "value2"
}
```

as well as

```js
[
  ["key", "value"]
  ["key", "value2"]
]
```

but notably does not allow

```js
{
  "key": ["value1", "value2"]
}
```

or

```js
[
  ["key", ["value1", "value2"]]
]
```

**Comment by cbiesinger** [comment]
@domenic  I'm curious why not `record<USVString, (USVString or sequence<USVString>)>`?

[FilePickerAcceptType](https://wicg.github.io/file-system-access/#dictdef-filepickeraccepttype) seems to use that.

**Comment by cbiesinger** [comment]
but I see that URLSearchParams uses your suggestion

**Comment by bvandersloot-mozilla** [comment]
Is there a reason to maintain the nonce when this is added?

**Comment by aaronpk** [comment]
No I don't think so. I mentioned in another issue (can't find it at the moment) that unless FedCM defines some specific processing behavior to use/validate the `nonce` parameter, it shouldn't be defined as a parameter in FedCM.

**Comment by samuelgoto** [comment]
> Is there a reason to maintain the nonce when this is added?

I don't think so. It would be a backwards incompatible change, so we'd have to coordinate with developers, which is doable but non-trivial. My plan is to get this in first, ship it in and once it settles start a migration and deprecation of `nonce`. Possible, but takes a lot of steps.

**Comment by aaronpk** [comment]
I just noticed the `param_` prefix in this proposal. I get the concern about not colliding with the FedCM-defined property names, but it does mean I need another step of preprocessing this request before I can hand off the request to my existing OAuth authorization endpoint.

It also feels a bit weird because `client_id` is a parameter already used in OAuth, but in order to use any other OAuth parameters I'd have to find them with the `param_` prefix in the post body request.

I'm not necessarily asking to change this, just wanted to get this written down while I'm thinking about it.

**Comment by panva** [comment]
> I just noticed the param_ prefix in this proposal.

Well this is a shocker, I toyed with the flagged (authz) predecessor implementation of this and didn't realize either. 

> I'm not necessarily asking to change this

I am ;)

> I get the concern about not colliding with the FedCM-defined property names

Solveable by having fedcm not use conflicting parameter names with known prior art if they happen to have a different meaning. We have known parameter public registries to support the process.

**Comment by npm1** [comment]
Yea I think ignoring any params conflicting with the FedCM specific ones could also work. Do any of these already conflict with some params commonly passed?

- client_id
- nonce
- account_id
- disclosure_text_shown
- is_auto_selected
- mode
- fields
- disclosure_shown_for

**Comment by aaronpk** [comment]
client_id is OAuth and OpenID Connect, and nonce is OpenID Connect, tho I believe the intent of these parameters is the same in FedCM and OAuth/OIDC so that shouldn't be a problem

**Comment by panva** [comment]
> Yea I think ignoring any params conflicting with the FedCM specific ones could also work. Do any of these already conflict with some params commonly passed?
> 
> * client_id
> * nonce
> * account_id
> * disclosure_text_shown
> * is_auto_selected
> * mode
> * fields
> * disclosure_shown_for

You can see the currently registered parameters in the [IANA registry](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#parameters).

client_id and nonce are registered oauth parameters, however, client_id has the same meaning so that's okay?, and nonce [should be removed](https://github.com/fedidcg/FedCM/issues/556#issuecomment-2122953704) (or at least made optional in the API during its deprecation) and also has the same meaning, although it is not applicable to non-OIDC flows so a removal is more appropriate.

**Comment by wparad** [comment]
I'd rather have a nested body (even one that had to be string parsed) compared to guessing which property keys have been abused.

**Comment by panva** [comment]
> I'd rather have a nested body (even one that had to be string parsed) compared to guessing which property keys have been abused.

Can you elaborate on "abused"? If the only param re-used by fedcm is client_id and further reuse is avoided by having a transparent development process which will look at prior art to avoid, what would a nested bag of attributes bring us? I believe that having existing oauth deployments need "just" a new response mode is a win.

**Comment by cbiesinger** [comment]
If we only disallow/ignore the known parameters, this is problematic from a compatibility perspective if we later add a new parameter.

We could disallow the existing parameters and reserve a prefix (like the `sec-` prefix for HTTP headers) for future parameters we will add.

But can you elaborate on why the param_ prefix is causing problems?

**Comment by cbiesinger** [comment]
In particular:
> Up until this there was a remote chance this API could directly be mapped to either the token_endpoint or the POST at authorization_endpoint (with a fedcm-specific response mode) if we chose to do so at our binding. With prefixes there isn't.

Couldn't that mapping strip the param_ prefix?

**Comment by aaronpk** [comment]
The `param_` prefix isn't a problem for net new implementations, but it is awkward to try to fit that into an existing OAuth authorization endpoint like @panva and I have done. It means more significant/awkward modification of the authorization endpoint to translate `param_x` to `x` before continuing with the rest of the existing processing.

**Comment by samuelgoto** [comment]
> I am ;) Up until this there was a remote chance this API could directly be mapped to either the token_endpoint or the POST at authorization_endpoint (with a fedcm-specific response mode) if we chose to do so at our binding. With prefixes there isn't.

The concern that I have with not having the prefixes is that (a) browser engineers are going to have to look into this [IANA Registry](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#parameters) **and** at every other identity protocol that exists today and in the future (e.g. SAML) every time we add one and (b) vice versa, this [IANA Registry](https://www.iana.org/assignments/oauth-parameters/oauth-parameters.xhtml#parameters) is going to have to look into parameters that the **browser** has added so that it doesn't collide. 

It is a two way street that can corner two layers of the stack that should be able to move otherwise independently.

I understand how the translation form `param_x` to `x` isn't the most ergonomic things in the world, but I think that, if you look at this holistically, it is the right trade-off.

**Comment by aaronpk** [comment]
Backwards compatibility issues aside, couldn't it go the other way around? FedCM parameters get a prefix, everything else is sent as is:

* `fedcm_account_id`
* `fedcm_fields`
* `fedcm_is_auto_selected`
* etc etc
* `client_id`
* `response_mode`
* `request_uri`
* etc etc

**Comment by cbiesinger** [comment]
but we can't leave backwards compatibility aside

**Comment by samuelgoto** [comment]
>  FedCM parameters get a prefix, everything else is sent as is:

We considered this too. 

Aside from backwards compatibility, I think our assumption is that the `id_assertion_endpoint` is a **FedCM** endpoint, not an OAuth one: the IdP **has** to write custom FedCM-specific code to handle it (e.g. check for CORS and `Sec-Fetch-Dest`): it speaks FedCM as input, and once it does, it can make translations easily.

**Comment by aaronpk** [comment]
That is how I implemented it in mine, which was to make a new endpoint with the FedCM processing (CORS, `Sec-Fetch-Dest`, origin check), then call the `generateAuthorizationCode` function that my OAuth authorization endpoint also calls.

**Comment by panva** [comment]
"What would an OpenID Connect binding for FedCM profile look like?" was the question posed at the IETF OAuth Interim meeting. Well, we have a shot at a clean binding with large chunk of existing deployments. A binding that actually fits the extension model of the OAuth protocol as well as the overall request flow of FedCM too. But the parameters must not get prefixed in order for that to be doable.

Adding new authorization-like endpoint that issues assertions is a much bigger undertaking than adding a [response mode](https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html#ResponseModes).

It is clear to me why you want to distinguish the params as well as that you don't want the burden of having to look at registries. I think you make a way bigger deal out of it than it actually is.

How about having a way to change what gets prefixed - either the fedcm parameters (explicit option) or the arbitrary idp ones (the default for the evergreen which is backwards compat).

There is valuable opportunity in making the request somewhat compatible with deployments of both OAuth and SAML. Not just maintained deployments but more importantly also ones that are beyond serviceability. Everything except the assertion endpoint's parameter handling can be done as a layer on top of a deployment which a) adds `Sec-Fetch-Dest` check, b) forwards the request **without mutating it** (so that any possible request signing is not affected), c) plucks the resulting artifacts out of the response and sends its own JSON one.

This could new breathe life to IdPs that would otherwise be dead weight after 3rd party cookie deprecation or whatever comes next.

**Comment by samuelgoto** [comment]
> Everything except the assertion endpoint's parameter handling can be done as a layer on top of a deployment which a) adds Sec-Fetch-Dest check, b) forwards the request without mutating it (so that any possible request signing is not affected), c) plucks the resulting artifacts out of the response and sends its own JSON one.

I think what's important is to maintain the architectural design choice of keeping the `id_assertion_endpoint` a FedCM endpoint, not an OAuth one. I think it is key to allow it evolve independently and autonomously, rather than in lock step with another protocol.

>  b) forwards the request without mutating it (so that any possible request signing is not affected), c) plucks the resulting artifacts out of the response and sends its own JSON one.

I think that should be an non-goal: making the FedCM request forwardable without mutations. If you run code that can intercept and forward, you can also mutate.

As I said, I think the `id_assertion_endpoint` is a FedCM endpoint, and making it be shared with another protocol is an unnecessary architectural design choice.

**Comment by bvandersloot-mozilla** [comment]
This is part of why I like the model of [the alternate operating mode of FedCM](https://github.com/fedidcg/CrossSiteCookieAccessCredential) more. There are no new server API endpoints dedicated to handling the browser API needed. Instead the RP can send a request to the IDP (which I would happily extend to include a Javascript dict) and the IDP can respond in its page, giving data back (which I would also happily extend) to the RP (and access to third party cookies if that is more convenient). 

I think the architecture of FedCM as is and the general naming of things, like nonce, client_id, and token really look from the outside like an identity protocol, rather than browser plumbing.

Also, this is a point where the form encoding is annoying- we can't include structure in the params because of it.

**Comment by bvandersloot-mozilla** [comment]
> but we can't leave backwards compatibility aside

Yes we _absolutely_ can :)

**Comment by panva** [comment]
> How about having a way to change what gets prefixed - either the fedcm parameters (explicit option) or the arbitrary idp ones (the default for the evergreen which is backwards compat).

@samuelgoto The above suggestion does not impede the architectural design choice, does not break backwards compatibility, and it makes integration with FedCM much easier for existing infrastructure.

> This is part of why I like the model of [the alternate operating mode of FedCM](https://github.com/fedidcg/CrossSiteCookieAccessCredential) more.

@bvandersloot-mozilla I was not aware of this and will give it a proper read in the coming week.

> I think the architecture of FedCM as is and the general naming of things, like nonce, client_id, and token really look from the outside like an identity protocol, rather than browser plumbing.

This a 💯.

> > but we can't leave backwards compatibility aside
> 
> Yes we _absolutely_ can :)

👍

**Comment by bc-pi** [comment]
I typically agree with @panva on just about everything but I'm not sure I share the same level of enthusiasm here.

Stepping back just a bit though, can we all maybe agree that the aesthetics of automatically prefixing `application/x-www-form-urlencoded` parameters with 'param_`  are not great and probably not the kind of thing that we'd want to enshrine "forever" in a standard?

**Comment by cbiesinger** [comment]
> Yes we absolutely can :)

I can't tell if I'm missing a joke but this is an API used by [0.9% of pageloads](https://chromestatus.com/metrics/feature/popularity#FederatedCredentialManagement) by a number of IdPs, so any breaking change needs to be carefully managed and communicated; certainly API owners would not let us just break this many pages even if we wanted to.

> Stepping back just a bit though, can we all maybe agree that the aesthetics of automatically prefixing `application/x-www-form-urlencoded` parameters with `param_` are not great and probably not the kind of thing that we'd want to enshrine "forever" in a standard?

I truly don't understand the problem?

(https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes is a somewhat related precedent)

**Comment by panva** [comment]
> I can't tell if I'm missing a joke but this is an API used by [0.9% of pageloads](https://chromestatus.com/metrics/feature/popularity#FederatedCredentialManagement) by a number of IdPs, so any breaking change needs to be carefully managed and communicated;

Isn't a substantial amount of these driven by GIS's own SDKs/provided scripts?

**Comment by samuelgoto** [comment]
> I can't tell if I'm missing a joke but this is an API used by 0.9% of pageloads by a number of IdPs, so any breaking change needs to be carefully managed and communicated; certainly API owners would not let us just break this many pages even if we wanted to.

Just to be concrete: it is possible to make backwards compatible changes, but they aren't free, so should have (really) good reasons to and do as few times as possible (e.g. batch them).

**Comment by aaronpk** [comment]
> (https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes is a somewhat related precedent)

The big difference here is the author of the HTML adds the `data-` prefix, rather than the browser adding it automatically.

**Comment by cbiesinger** [comment]
> > I can't tell if I'm missing a joke but this is an API used by [0.9% of pageloads](https://chromestatus.com/metrics/feature/popularity#FederatedCredentialManagement) by a number of IdPs, so any breaking change needs to be carefully managed and communicated;
> 
> Isn't a substantial amount of these driven by GIS's own SDKs/provided scripts?

Even for GIS changes are not free, and I'm more concerned about other IDPs who may get scared off by FedCM making too many breaking changes. That said -- as Sam said, there are probably still few enough IdPs that we can make changes if we have good reasons and manage the rollout.


> > (https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes is a somewhat related precedent)
> 
> The big difference here is the author of the HTML adds the `data-` prefix, rather than the browser adding it automatically.

We could certainly require the caller of FedCM to add a param or data prefix if that were helpful?

**Comment by panva** [comment]
> We could certainly require the caller of FedCM to add a param or data prefix if that were helpful?

Or this?

> How about having a way to change what gets prefixed - either the fedcm parameters (explicit option) or the arbitrary idp ones (the default for the evergreen which is backwards compat).

**Comment by samuelgoto** [comment]
To get back to my original point, aside of backwards compatibility, which, as I said, is [possible but not free](https://github.com/fedidcg/FedCM/issues/556#issuecomment-2135670331):

> @samuelgoto The above suggestion does not impede the architectural design choice, does not break backwards compatibility, and it makes integration with FedCM much easier for existing infrastructure.

The point that I'm trying to make is that the `id_assertion_endpoint` is a FedCM one, not an OAuth one: it is not a design goal to make it backwards compatible to OAuth without any modification.

There is a series of key privacy and security checks (e.g. the `Sec-Fetch-Dest` header, the `account_id` that was selected, whether the account was [`is_auto_selected`](https://fedidcg.github.io/FedCM/#fetch-identity-assertion), whether the `disclosure_text_shown`, etc) that need to be made in the `id_assertion_endpoint`, and we want to reserve the right to evolve it autonomously and independently.

Security first, ergonomics second, IMHO.

**Comment by bvandersloot-mozilla** [comment]
> I can't tell if I'm missing a joke but this is an API used by [0.9% of pageloads](https://chromestatus.com/metrics/feature/popularity#FederatedCredentialManagement) by a number of IdPs, so any breaking change needs to be carefully managed and communicated; certainly API owners would not let us just break this many pages even if we wanted to.

No joking here. Managing backwards incompatible changes is part of the challenge you accept when shipping to the general population before

**Comment by samuelgoto** [comment]
> Managing backwards incompatible changes is part of the challenge of shipping an experimental API that doesn't have the implementation experience, standard set, or another browser shipping

Yeah, I think we are roughly on the same page: it is a risk that we are happy that we took, and are happy to manage it now.

Part of managing that risk is knowing that it is both likely necessary and but also not free.

LMK if you disagree with anything in this response here: https://github.com/fedidcg/FedCM/issues/556#issuecomment-2135670331

**Comment by panva** [comment]
> The point that I'm trying to make is that the id_assertion_endpoint is a FedCM one, not an OAuth one: it is not a design goal to make it backwards compatible to OAuth without any modification. 

Yes, I do believe the circumstance that there stands to be a chance of a clean oauth profile for fedcm and other existing prior idp art is entirely accidental.

> I think it is unsafe to suggest so

I agree it's unsafe, but also noone suggested it. An oauth binding for fedcm would't dare omitting the mandatory idp provisions set forth by fedcm regardless of whether its parameters are prefixed or not. 

> one has to process a FedCM request intentionally, rather than by accident.

100% agree. But one can intentionally (or unintentionally) mess up fedcm's provisions regardless of how custom vs fedcm-specific params are transmitted. Bad implementations are inevitable. They are also not a strong argument against an affordance for respecting years of prior idp art.

**Comment by samuelgoto** [comment]
> They are also not a strong argument against an affordance for respecting years of prior idp art.

I think we are starting to talk past each other now: I'm not suggesting that we ignore prior art, I'm suggesting that this is an entirely new layer that's being introduced, and that the parameters are specific to that layer. I already made this point, so don't think it is a matter of clarification.

**Comment by timcappalli** [comment]
Proposal on the 2024-06-11 call: what if this was POST w/ JSON payload vs parameters? Then the `params_` prefix could be dropped from the key names.

**Comment by aaronpk** [comment]
Clarifying question for FedCM implementations and intent going forward... are the arbitrary `params` values only ever strings or could they be other JSON data types including complex objects? Does the browser care? e.g.

```
let {token} = await navigator.credentials.get({
  identity: {
    providers: [{
      clientId: "1234",
      nonce: "234234",
      loginHint: "previous@user.com",
      configURL: "https://idp.example/fedcm.json",
      params: {
        "response_type": "id_token",
        "scope": "photos:read photos:write",
        "foo": true,
        "authorization_details": [
           {
             "type": "payment_initiation", 
             "amount": {"currency": "EUR", "amount": 123.5} 
           }
        ]
      }
    },
  }
});
```

**Comment by cbiesinger** [comment]
The values are currently defined as strings in the IDL, so if you try to pass an object, you will get standard JS behavior to convert to string, which will likely result in "[Object object]" or somesuch. (you can of course call JSON.stringify yourself)



> Proposal on the 2024-06-11 call: what if this was POST w/ JSON payload vs parameters? Then the `params_` prefix could be dropped from the key names.

As a note, because of how CORS works (https://fetch.spec.whatwg.org/#cors-safelisted-request-header) this would make the ID assertion fetch require a CORS preflight request. This isn't necessarily a problem but I wanted to point that out.

**Comment by bvandersloot-mozilla** [comment]
It seems reasonable to support arbitrary structure. That is what the discussion in #578 led to. But things get weird when you start having numbers that parse out of JSON imprecisely, like `JSON.parse(12345678901234567890) == 12345678901234567000`.

**Comment by cbiesinger** [comment]
Hmm if we start supporting arbitrary JSON objects and (presumably) serialize that into a string to send as a parameter value -- maybe we should go a step further and just send `params=${JSON.stringify(identity.providers[N].params)`?

**Comment by samuelgoto** [comment]
> params=${JSON.stringify(identity.providers[N].params)

Yeah, that occurred to me, and it seems right. That's what we are going to for the digital credentials API.

**Comment by cbiesinger** [comment]
Adding agenda+ to discuss the proposal in the last two comments & to hear if there was an outcome at the IETF meeting

**Comment by elf-pavlik** [comment]
I understand that the discussion is about https://www.w3.org/TR/fedcm/#idp-api-id-assertion-endpoint

Which currently states:

> It will also contain the following parameters in the request body `application/x-www-form-urlencoded`

Reading the proposal

> Hmm if we start supporting arbitrary JSON objects and (presumably) serialize that into a string to send as a parameter value -- maybe we should go a step further and just send `params=${JSON.stringify(identity.providers[N].params)`?

I don't understand why the content type of the request body can't be changed to `application/json`

I'm sorry if I missed something obvious.

**Comment by cbiesinger** [comment]
We discussed this at today's FedID CG call. The outcome was that everyone is happier now that we no longer prefix parameters and support arbitrary values (not just strings). There was no particular preference for application/json vs urlencoded as far as I can tell.

**Comment by npm1** [comment]
I suppose if there are no benefits to using application/json then we can keep the current as it a) avoids backwards incompatible change and b) avoids CORS preflights (which would need to be supported by IDPs, although if needed I don't think it would be a big deal)

**Comment by samuelgoto** [comment]
> I don't understand why the content type of the request body can't be changed to application/json

I think this has its own merits and challenges on its own, outside of the discussion of how this issue. I kicked off https://github.com/w3c-fedid/FedCM/issues/644 to decouple the discussion, so that we can make forward progress on this issue independently, and move to `text/json` as a separate step.
